lm_eval/tasks/sat.py

```diff
@@ -94,5 +97,20 @@ def doc_to_text(self, doc, include_target=True):
         return text
 
     def evaluate(self, docs, lm):
-        # TODO: Write evaluation function
-        raise NotImplementedError()
+        golds = [doc["answer_key"] for doc in docs]
+        preds = []
+        for doc in tqdm_lib.tqdm(docs):
+            ctx = self.fewshot_context(
+                doc=doc,
+                num_fewshot=1,
+                provide_description=None,
+                # unless Dataset evaluate()s should get num_fewshot/ provide_description
+            )
+            probs_before_numpy = []
+            for choice in doc["choices"]:
+                this_choice = " " + choice
+                probs_before_numpy.append(lm.loglikelihood(ctx, this_choice))
+            probs = np.array(probs_before_numpy)
+            preds.append(np.argmax(probs))
+
+        return simple_accuracy_metric(preds=preds, golds=golds)

```

Comment:
We're switching to the new evaluation framework. Check out the interface in `base.py` and the example BoolQ implementation in `superglue.py`, that's what you need to implement.

Reply:
@leogao2 So instead of `evaluate()`, I get the results already and build a `process_results()` method?

Reply:
The functions you need to implement are `doc_to_target`, `construct_requests`, and `process_results`.

`doc_to_target` is basically like `doc_to_text` but for the answer you want the lm to predict for a document.
`construct_requests` returns a list/tuple of `Request` objects (constructed using a `RequestFactory`) that represent things you want to query the LM with. You can think of this sort of like an inside-out `sess.run`, each `Request` you return is like an argument you give to `sess.run`.
`process_results` is called after the LM has been run and gives you back the results of whatever you asked for in `construct_requests`. 

You can check out BoolQ for an example of how this is all done, it'll probably be a lot clearer after looking at the example.

Reply:
Ah, thanks for clearing all this up! I'll get on that